src/security/acl_api.c

```diff
@@ -0,0 +1,450 @@
+		struct daos_ace *new_ace)
+{
+	struct daos_ace	*current;
+	uint8_t		*pen;
+	bool		new_written = false;
+
+	current = daos_acl_get_first_ace(acl);
+	pen = new_acl->dal_ace;
+	while (current != NULL) {
+		if (!new_written && current->dae_principal_type >
+				new_ace->dae_principal_type)  {
+			new_written = true;
+			pen = write_ace(new_ace, pen);
+		} else if (!new_written && principals_match(current,
+					new_ace)) {
+			new_written = true;
+			pen = write_ace(new_ace, pen);
+
+			/* new ACE already in old list - shorten the length */
+			new_acl->dal_len = acl->dal_len;
+
+			current = daos_acl_get_next_ace(acl, current);
+		} else {
+			pen = write_ace(current, pen);
+			current = daos_acl_get_next_ace(acl, current);
+		}
+	}
+
+	/* didn't insert it - tack it on the end */
+	if (!new_written) {
+		write_ace(new_ace, pen);
+	}
+}
+
+int
+daos_acl_add_ace_realloc(struct daos_acl *acl, struct daos_ace *new_ace,
+		struct daos_acl **new_acl)
+{
+	int		new_ace_len;
+	int		new_len;
+
+	if (acl == NULL || new_acl == NULL) {
+		return -DER_INVAL;
+	}
+
+	new_ace_len = daos_ace_get_size(new_ace);
+	if (new_ace_len < 0) {
+		/* ACE was invalid */
+		return -DER_INVAL;
+	}
+
+	new_len = acl->dal_len + new_ace_len;
+
+	D_ALLOC(*new_acl, sizeof(struct daos_acl) + new_len);
+	if (*new_acl == NULL) {
+		return -DER_NOMEM;
+	}
+
+	(*new_acl)->dal_ver = acl->dal_ver;
+	(*new_acl)->dal_len = new_len;
+
+	copy_acl_with_new_ace_inserted(acl, *new_acl, new_ace);
+
+	return DER_SUCCESS;
+}
+
+static bool
+type_is_valid(enum daos_acl_principal_type type)
+{
+	bool result = false;
+
+	switch (type) {
+	case DAOS_ACL_USER:
+	case DAOS_ACL_GROUP:
+	case DAOS_ACL_OWNER:
+	case DAOS_ACL_OWNER_GROUP:
+	case DAOS_ACL_EVERYONE:
+		result = true;
+		break;
+	}
+
+	return result;
+}
+
+static bool
+type_needs_name(enum daos_acl_principal_type type)
+{
+	/*
+	 * The only ACE types that require a name are User and Group. All others
+	 * are "special" ACEs that apply to an abstract category.
+	 */
+	if (type == DAOS_ACL_USER || type == DAOS_ACL_GROUP) {
+		return true;
+	}
+
+	return false;
+}
+
+static bool
+principal_meets_type_requirements(enum daos_acl_principal_type type,
+		const char *principal_name, size_t principal_name_len)
+{
+	return	(!type_needs_name(type) ||
+		(principal_name != NULL && principal_name_len != 0));
+}
+
+int
+daos_acl_remove_ace_realloc(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal_name,
+		size_t principal_name_len, struct daos_acl **new_acl)
+{
+	struct daos_ace	*current;
+	uint8_t		*pen;
+
+	if (acl == NULL || new_acl == NULL || !type_is_valid(type) ||
+		!principal_meets_type_requirements(type, principal_name,
+				principal_name_len)) {
+		return -DER_INVAL;
+	}
+
+	if (daos_acl_get_ace_for_principal(acl, type, principal_name) == NULL) {
+		/* requested principal not in the list */
+		return -DER_NONEXIST;
+	}
+
+	D_ALLOC(*new_acl, sizeof(struct daos_acl) + acl->dal_len);
```

Comment:
Wont this allocate more space than necessary? Again only once ACEs worth but it is larger than we need.

Reply:
Yep. As above, we could do the extra work to calculate the actual length before we start. I guess the question is which is cheaper in this case - compute cycles or memory.

Reply:
Same comment as above. In this case we actually already check whether the principal is already there - we can absolutely use it to calculate the real length.